cpp/src/reductions/reduction_operators.cuh

```diff
@@ -0,0 +1,260 @@
+
+    template<typename T>
+    __device__
+    T operator() (const T &lhs, const T &rhs) {
+        return lhs <= rhs? lhs: rhs;
+    }
+
+    template<typename T>
+    static constexpr T identity() { return std::numeric_limits<T>::max(); }
+};
+
+struct DeviceMax{
+    typedef IdentityLoader Loader;
+
+    template<typename T>
+    __device__
+    T operator() (const T &lhs, const T &rhs) {
+        return lhs >= rhs? lhs: rhs;
+    }
+
+    template<typename T>
+    static constexpr T identity() { return std::numeric_limits<T>::lowest(); }
+};
+
+
+// ------------------------------------------------------------------------
+
+template <typename T, typename Op>
+__forceinline__  __device__
+void genericAtomicOperation(T& existing_value, T const & update_value, Op op)
+{
+  T old_value = existing_value;
+  T expected{old_value};
+
+  // Attempt to perform the aggregation with existing_value and
+  // store the result atomically
+  do
+  {
+    expected = old_value;
+
+    const T new_value = op(update_value, old_value);
+
+    old_value = genericAtomicCAS(&existing_value, expected, new_value);
+  }
+  // Guard against another thread's update to existing_value
+  while( expected != old_value );
+}
+
+// ------------------------------------------------------------------------
+// specialized functions for operators
+// `atomicAdd` supports int32, float, double (signed int64 is not supproted.)
+// `atomicMin`, `atomicMax` support int32_t, int64_t
+
+template <>
+__forceinline__  __device__
+void genericAtomicOperation(
+    int32_t& existing_value, int32_t const & update_value, DeviceSum op)
+{
+    atomicAdd(&existing_value, update_value);
+}
+
+template <>
+__forceinline__  __device__
+void genericAtomicOperation(
+    float& existing_value, float const & update_value, DeviceSum op)
+{
+    atomicAdd(&existing_value, update_value);
+}
+
+template <>
+__forceinline__  __device__
+void genericAtomicOperation(
+    double& existing_value, double const & update_value, DeviceSum op)
+{
+    atomicAdd(&existing_value, update_value);
+}
+
+template <>
+__forceinline__  __device__
+void genericAtomicOperation(
+    int32_t& existing_value, int32_t const & update_value, DeviceMin op)
+{
+    atomicMin(&existing_value, update_value);
+}
+
+template <>
+__forceinline__  __device__
+void genericAtomicOperation(
+    int64_t& existing_value, int64_t const & update_value, DeviceMin op)
+{
+    atomicMin(reinterpret_cast<long long*>(&existing_value),
+        static_cast<long long>(update_value));
+}
+
+template <>
+__forceinline__  __device__
+void genericAtomicOperation(
+    int32_t& existing_value, int32_t const & update_value, DeviceMax op)
+{
+    atomicMax(&existing_value, update_value);
+}
+
+template <>
+__forceinline__  __device__
+void genericAtomicOperation(
+    int64_t& existing_value, int64_t const & update_value, DeviceMax op)
+{
+    atomicMax(reinterpret_cast<long long*>(&existing_value),
+        static_cast<long long>(update_value));
+}
+
```

Comment:
Can you add specializations for the `wrapper` types as well? https://github.com/rapidsai/cudf/blob/branch-0.6/cpp/src/utilities/wrapper_types.hpp#L55

These should perform the appropriate atomic operation on the `wrapper`s underlying value.

Reply:
The input typename T for `genericAtomicOperation` is already underlying type.
The underlying type is used before launching the kernel.

https://github.com/rapidsai/cudf/blob/e68dc4353db5c774b0c3a9bc651a4f4822910306/cpp/src/reductions/reductions.cu#L112-L119

Reply:
Repasting my comment here:

We really don't want to do that, the underlying type should not be used directly here. It should only be used inside wrapper operators and specializations of atomic* operations.

The reason for this is because some of the wrapper types may have different semantics than their underlying type. Perfect example is cudf::bool8 uses a int8_t as the underlying type. However cudf::bool8{1} + cudf::bool8{1} == 1, if you were to use the underlying type directly, you would produce the incorrect results of 1 + 1 = 2.

We want to control the `wrapper` types at the operator level, but by using the underlying type directly, you've made it so that we can't add a specialization/overload for `atomicAdd` on `cudf::bool8` for example. 

You should think of the wrapper types as opaque objects that you should not be accessing the underlying type nor value.

The only entities that should know about the underlying type or value are wrapper operators and other operators that are specialized to work on the wrapper types, e.g., atomic* operations.

Reply:
OK, I will avoid to use UnderlyingType except for atimic oprations.

Btw, I don't think cudf::bool8 is a countable type. So, `reduction` (including min, max) won't support `cudf::bool8`.
Is it good?

Reply:
> Btw, I don't think cudf::bool8 is a countable type. So, `reduction` (including min, max) won't support `cudf::bool8`.

We'd need to check what Pandas does. You technically _could_ do `min`, `max`, `sum` across a boolean column... it just wouldn't be that useful.